### ISSUE TITLE
Raw web keyCode cannot be extracted from KeyEvent

### ISSUE THREAD

**Issue by Senne021** [flutter/flutter#147162]
### Steps to reproduce

We have migrated our keyboard handling from Flutter 3.16 to 3.19 according to [this migration guide](https://docs.flutter.dev/release/breaking-changes/key-event-migration#migrating-your-code-that-uses-rawkeyevent). After the migration we can't get the keyCode from the KeyEvent as we did before.
We're using Flutter for a webos application and for that we need to get the keyCode from the `RawKeyEventDataWeb` to distinguish between the special keys coming from the [LG magic remote](https://webostv.developer.lge.com/develop/guides/magic-remote).

### Expected results

KeyCode like 413 for stop

### Actual results

We're getting the `usbHidUsage`: `0x01700000000` and the `LogicalKeyboardKey` `unidentified`

### Code sample

<details open><summary>Code sample</summary>

```dart
Focus(
      onKeyEvent: (FocusNode node, KeyEvent event) {
        if (event is RawKeyDownEvent && event.data is RawKeyEventDataWeb) {
          final keyCode = (event.data as RawKeyEventDataWeb).keyCode;
          // Do something with the key code
        }
        return KeyEventResult.ignored;
      },
      child: ...

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.19.6, on macOS 14.4.1 23E224 darwin-arm64, locale de-DE)
    • Flutter version 3.19.6 on channel stable at ...
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 54e66469a9 (5 days ago), 2024-04-17 13:08:03 -0700
    • Engine revision c4cd48e186
    • Dart version 3.3.4
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at ...
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A240d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio2.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.88.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 14.4.1 23E224 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 124.0.6367.62

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

**Comment by Senne021** [comment]
@darshankawar Thank you for your reply.

> Does it mean for LG Smart TV ? and you using the LG smart remote to navigate flutter app on it ? What's the specific keys are you pressing ?

Yes, we're using our app on LG Smart TVs. The missing key code occurs on both the real device and the simulator. Only the media and back buttons are affected by this issue.

> Can you provide us the entire log ?

Here is the entire log of different key events:

Play and stop on the LG remote:
```
    KeyUpEvent(
        physicalKey: 
          PhysicalKeyboardKey(
              usbHidUsage: "0x1700000000",
              debugName: null
          ), 
        logicalKey: 
          LogicalKeyboardKey(
              keyId: "0x100000001", 
              keyLabel: "Unidentified", 
              debugName: null
          ), 
        character: null, 
        timeStamp: 0:01:03.819099
    )
```

And the key event of one of the arrow keys that work on an LG remote control:
```
    KeyDownEvent(
        physicalKey: PhysicalKeyboardKey(
            usbHidUsage: "0x00070051", 
            debugName: "Arrow Down"
        ), logicalKey: 
        LogicalKeyboardKey(
            keyId: "0x100000301", 
            keyLabel: "Arrow Down", 
            debugName: "Arrow Down"
        ), 
        character: null, 
        timeStamp: 0:00:14.981599
    )
```

**Comment by darshankawar** [comment]
Thanks for the update. Due to lack of said emulator / device, we can't currently verify this. Keeping the issue open for team's input / attention / tracking.

**Comment by justinmc** [comment]
CC @gspencergoog I believe you wrote the feature mentioned in the description here.

**Comment by gspencergoog** [comment]
Yes, but the `RawKeyEvent` implementation has been deprecated.  The replacement `KeyEvent` implementation doesn't support access to the platform specific keycodes anymore, so that the entire key event system is more platform agnostic. Ideally, the fix for this is to create an appropriate mapping for the missing keys so that they are identifiable in the same way as other keys are, and the application doesn't need to check web keycodes at all.

**Comment by Senne021** [comment]
@gspencergoog Thank you for your reply. How can I add this type of mapping to Flutter? Do you have an example of that?

**Comment by gspencergoog** [comment]
Sorry, just to be more clear: the mapping isn't something you'd do normally when making an app, it's a base mapping in Flutter itself.

This is controlled by the keycode generator in https://github.com/flutter/flutter/tree/master/dev/tools/gen_keycodes and what would need to happen is an addition to the keycode database that has the correct mappings, whatever those are. You're welcome to try and build a PR for that yourself, otherwise someone on the Flutter team will need to do it.

**Comment by yjbanov** [comment]
@gspencergoog Do you think that expanding the mappings is a scalable solution? Would it make sense for us to provide hooks for extending the mappings in separate packages? For example, there could be a `package:webos` that adds WebOS-specific functionality.

**Comment by gspencergoog** [comment]
It can already be done in a scalable way: you can define your own `LogicalKeyboardKey`s, and `PhysicalKeyboardKey`s in a package if you want.  The mappings just generate the "standard" ones that come with Flutter to contribute to the common set for everyone to use.

However, in this case, the mapping seems to for a specific key called `LogicalKeyboardKey.unidentified` in the [ mappings](https://github.com/gspencergoog/flutter/blob/b9caea7c2ca87782c4abddbded3051f5f266232b/packages/flutter/lib/src/services/keyboard_maps.g.dart#L2466-L2467), and I'm not quite sure what that's indicating.  Perhaps there is something on the engine side that needs to be added?  @dkwingsmt do you know?

**Comment by yjbanov** [comment]
I see, and wow, that's a big file! I wonder if we should break it up by platform and lazy-load only the mappings needed by the current platform. But that's a different story.

**Comment by gspencergoog** [comment]
The mappings are converted into [Dart code](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/keyboard_key.g.dart), so we don't load it or include it with the app or anything, it's strictly there for codegen purposes (and codegen only happens we we update the mapping and the results are checked in). The unused dart symbols are elided at compile time, so size isn't an issue there either.

**Comment by Senne021** [comment]
@gspencergoog Thx for your input. 
I've have looked a little bit deeper into the Flutter keycode generator you mentioned and found a problem when adding the webos keycode. The logical and physical keys for web applications are mapped according to the [`code`](https://github.com/flutter/flutter/blob/ddd6c8695084d6661574bf941b79e2eca252be58/packages/flutter/lib/src/services/raw_keyboard_web.dart#L104) and not the `keycode` (which doesn't seem to be used at all). When a key is pressed on a webos remote, an event is triggered where only the keyCode is set. The code value is empty and therefore can't be used for mapping. 
Do we need web [scan codes](https://github.com/flutter/flutter/blob/ddd6c8695084d6661574bf941b79e2eca252be58/dev/tools/gen_keycodes/data/physical_key_data.g.json#L3277) as some kind of fallback if the `code` value can't be found in the mapping?

**Comment by simonisinger** [comment]
Any movement here? this problem dont only affects web or webos but also android tv and fire tv. Its working wonderful with RawKeyboardListener and the known key codes. I have implemented it a few years ago for it. either make the new randomized key system optional and add the old keycodes again or deprecate the new system and remove the deprecation on the old.

The RawKeyboardListener is very flexible and isnt bound to one type of device and the keyboardlistener want only devices with pc keyboards.
